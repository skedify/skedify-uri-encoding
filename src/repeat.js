export default function repeat(string, count) {
  if (string.repeat && typeof string.repeat === 'function') {
    return string.repeat(count);
  }

  let counter = Number(count);
  let newString = string;

  if (isNaN(counter)) {
    counter = 0;
  }

  if (counter < 0) {
    throw new RangeError('repeat count must be non-negative');
  }

  if (counter === Infinity) {
    throw new RangeError('repeat count must be less than infinity');
  }

  counter = Math.floor(counter);

  if (newString.length === 0 || counter === 0) {
    return '';
  }

  // Ensuring count is a 31-bit integer allows us to heavily optimize the
  // main part. But anyway, most current (August 2014) browsers can't handle
  // strings 1 << 28 chars or longer, so:
  if (newString.length * counter >= 1 << 28) {
    throw new RangeError('repeat count must not overflow maximum string size');
  }

  let rpt = '';

  /* eslint-disable-next-line no-constant-condition */
  while (true) {
    /* eslint-disable-next-line no-bitwise */
    if ((counter & 1) === 1) {
      rpt += newString;
    }

    /* eslint-disable-next-line no-bitwise */
    counter >>>= 1;

    if (counter === 0) {
      break;
    }

    newString += newString;
  }

  return rpt;
}
