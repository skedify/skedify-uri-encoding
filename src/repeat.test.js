import repeat from './repeat'

const createStringWithoutRepeat = string => {
  const aString = new String(string)
  aString.repeat = undefined

  return aString
}

const testCases = [
  { string: 'abc', toBe: '' },
  { string: 'abc', count: undefined, toBe: '' },
  { string: 'abc', count: null, toBe: '' },
  { string: 'abc', count: false, toBe: '' },
  { string: 'abc', count: NaN, toBe: '' },
  { string: 'abc', count: 'abc', toBe: '' },
  { string: 'abc', count: -0, toBe: '' },
  { string: 'abc', count: +0, toBe: '' },
  { string: 'abc', count: 1, toBe: 'abc' },
  { string: 'abc', count: 2, toBe: 'abcabc' },
  { string: 'abc', count: 3, toBe: 'abcabcabc' },
  { string: 'abc', count: 4, toBe: 'abcabcabcabc' },
]

describe('repeat (when String.repeat exist)', () => {
  testCases.forEach(testCase => {
    it(`should return ${testCase.toBe} on string ${
      testCase.string
    } with count ${testCase.count}`, () => {
      expect(repeat(testCase.string, testCase.count)).toBe(testCase.toBe)
    })
  })
})

describe('repeat (when String.repeat does not exist)', () => {
  testCases.forEach(testCase => {
    it(`should return ${testCase.toBe} on string ${
      testCase.string
    } with count ${testCase.count}`, () => {
      expect(
        repeat(createStringWithoutRepeat(testCase.string), testCase.count)
      ).toBe(testCase.toBe)
    })
  })

  it('should throw a range error when count is negative', () => {
    expect(() => {
      repeat(createStringWithoutRepeat('abc'), -1)
    }).toThrowErrorMatchingSnapshot()
  })

  it('should throw a range error when count is negative infinity', () => {
    expect(() => {
      repeat(createStringWithoutRepeat('abc'), -Infinity)
    }).toThrowErrorMatchingSnapshot()
  })

  it('should throw a range error when count is infinity', () => {
    expect(() => {
      repeat(createStringWithoutRepeat('abc'), Number(Infinity))
    }).toThrowErrorMatchingSnapshot()
  })

  it('should throw an exception when strings are 1 << 28 chars or longer', () => {
    expect(() => {
      repeat(createStringWithoutRepeat('a'), Math.pow(2, 28))
    }).toThrowErrorMatchingSnapshot()
  })
})
