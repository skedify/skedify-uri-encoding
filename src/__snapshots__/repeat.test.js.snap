// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`repeat (when String.repeat does not exist) should throw a range error when count is infinity 1`] = `"repeat count must be less than infinity"`;

exports[`repeat (when String.repeat does not exist) should throw a range error when count is negative 1`] = `"repeat count must be non-negative"`;

exports[`repeat (when String.repeat does not exist) should throw a range error when count is negative infinity 1`] = `"repeat count must be non-negative"`;

exports[`repeat (when String.repeat does not exist) should throw an exception when strings are 1 << 28 chars or longer 1`] = `"repeat count must not overflow maximum string size"`;
